src/PrestaShopBundle/EventListener/ModuleActivatedListener.php

```diff
@@ -0,0 +1,198 @@
+ * needs please refer to http://www.prestashop.com for more information.
+ *
+ * @author    PrestaShop SA <contact@prestashop.com>
+ * @copyright 2007-2018 PrestaShop SA
+ * @license   https://opensource.org/licenses/OSL-3.0 Open Software License (OSL 3.0)
+ * International Registered Trademark & Property of PrestaShop SA
+ */
+
+namespace PrestaShopBundle\EventListener;
+
+use Doctrine\Common\Annotations\AnnotationException;
+use PrestaShopBundle\Security\Annotation\ModuleActivated;
+use Symfony\Component\HttpFoundation\Session\SessionInterface;
+use Symfony\Component\HttpKernel\Event\FilterControllerEvent;
+use Symfony\Component\Translation\TranslatorInterface;
+use Symfony\Component\HttpFoundation\RedirectResponse;
+use Symfony\Component\Routing\RouterInterface;
+use Doctrine\Common\Annotations\Reader;
+use Doctrine\Common\Util\ClassUtils;
+use ReflectionObject;
+use ReflectionClass;
+
+/**
+ * Allow a redirection to the right url when using ModuleActivated annotation
+ * and the module is inactive.
+ */
+class ModuleActivatedListener
+{
+    /**
+     * @var RouterInterface
+     */
+    private $router;
+
+    /**
+     * @var TranslatorInterface
+     */
+    private $translator;
+
+    /**
+     * @var SessionInterface
+     */
+    private $session;
+
+    /**
+     * @var Reader
+     */
+    private $annotationReader;
+
+    /**
+     * ModuleActivatedListener constructor.
+     *
+     * @param RouterInterface     $router
+     * @param TranslatorInterface $translator
+     * @param SessionInterface    $session
+     * @param Reader              $annotationReader
+     */
+    public function __construct(
+        RouterInterface $router,
+        TranslatorInterface $translator,
+        SessionInterface $session,
+        Reader $annotationReader
+    ) {
+        $this->router = $router;
+        $this->translator = $translator;
+        $this->session = $session;
+        $this->annotationReader = $annotationReader;
+    }
+
+    /**
+     * @param FilterControllerEvent $event
+     *
+     * @throws AnnotationException
+     * @throws \ReflectionException
+     */
+    public function onKernelController(FilterControllerEvent $event)
+    {
+        if (!$event->isMasterRequest()) {
+            return;
+        }
+
+        $controller = $event->getController();
+
+        if (!is_array($controller)) {
+            return;
+        }
+
+        list($controllerObject, $methodName) = $controller;
+        $moduleActivated = $this->getAnnotation($controllerObject, $methodName);
+
+        if (null === $moduleActivated) {
+            return;
+        }
+
+        /** @var \Module $module */
+        $module = \Module::getInstanceByName($moduleActivated->getModuleName());
+        if (null === $module) {
+            return;
+        }
+
+        if (!$module->active) {
+            $this->showNotificationMessage($moduleActivated);
+            $url = $this->router->generate($moduleActivated->getRedirectRoute());
```

Comment:
also, I'm not aware of this public method

Reply:
which method are you talking about? generate?

Reply:
getRedirectRoute(), I can't find it in the module class.

Reply:
it's not in the module class, it's in the `ModuleActivated` class which is the Annotation class

Reply:
ok :)